pkg/rng/entropy.go

```diff
@@ -0,0 +1,164 @@
+	// in seconds
+	EntropyFeedTime time.Duration = 10
+	// RandomDevice is the linux random device
+	RandomDevice = "/dev/random"
+	// HwRandomDevice is the linux hw random device
+	HwRandomDevice = "/dev/hwrng"
+)
+
+// Can be extended but keep in mind to priorize
+// more secure random sources like hw random over
+// timer, jitter based mechanisms. Zero is the highest
+// priority.
+// <rng-name> : <priority>
+var trngList = map[string]int{
+	"tpm-rng":        0,
+	"intel-rng":      1,
+	"amd-rng":        1,
+	"timeriomem-rng": 2,
+}
+
+// Searches for available True Random Number Generator
+// inside the kernel api and sets the most secure on if
+// available which seeds /dev/hwrng
+func setAvailableTRNG() (bool, error) {
+	var (
+		currentRNG    string
+		availableRNGs []string
+		rngs          []string
+	)
+
+	currentFileData, err := ioutil.ReadFile(HwRandomCurrentFile)
+	if err != nil {
+		return false, err
+	}
+	currentRNG = string(currentFileData)
+
+	availableFileData, err := ioutil.ReadFile(HwRandomAvailableFile)
+	if err != nil {
+		return false, err
+	}
+	availableRNGs = strings.Split(string(availableFileData), " ")
+
+	for key, value := range trngList {
+		if key == currentRNG && value == 0 {
+			return true, nil
+		}
+	}
+
+	for _, rng := range availableRNGs {
+		for key := range trngList {
+			if rng == key {
+				rngs = append(rngs, key)
+			}
+		}
+	}
+
+	if len(rngs) <= 0 {
+		return false, nil
+	}
+
+	if err = ioutil.WriteFile(HwRandomCurrentFile, []byte(rngs[0]), 0644); err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+// UpdateLinuxRandomness seeds random data from
+// /dev/hwrng into /dev/random based on a timer and
+// the entropy pool size
+// Usage:
+// go UpdateLinuxRandomness()
+func UpdateLinuxRandomness() error {
+	good, err := setAvailableTRNG()
+	if err != nil {
+		return err
+	}
+	if !good {
+		return errors.New("Could not find a good TRNG")
+	}
+
+	randomPoolSizeData, err := ioutil.ReadFile(RandomPoolSizeFile)
+	if err != nil {
+		return err
+	}
+
+	formatted := strings.TrimSuffix(string(randomPoolSizeData), "\n")
+	randomPoolSize, err := strconv.ParseUint(formatted, 10, 32)
+	if err != nil {
+		return err
+	}
+
+	hwRng, err := os.OpenFile(HwRandomDevice, os.O_RDONLY, os.ModeDevice)
+	if err != nil {
+		return err
+	}
+
+	defer hwRng.Close()
+
+	rng, err := os.OpenFile(RandomDevice, os.O_APPEND|os.O_WRONLY, os.ModeDevice)
+	if err != nil {
+		return err
+	}
+
+	defer rng.Close()
+
+	for {
+		randomEntropyAvailableData, err := ioutil.ReadFile(RandomEntropyAvailableFile)
+		if err != nil {
+			return err
+		}
+
+		formatted := strings.TrimSuffix(string(randomEntropyAvailableData), "\n")
+		randomEntropyAvailable, err := strconv.ParseUint(formatted, 10, 32)
+		if err != nil {
+			return err
+		}
+
+		randomBytesNeeded := randomPoolSize - randomEntropyAvailable
+		if randomBytesNeeded > 0 {
+			var random = make([]byte, randomBytesNeeded)
+			if _, err = hwRng.Read(random); err != nil {
+				return err
```

Comment:
We might want to also do something here if we can't read many bytes from the device. What happens if we get 0 bytes every time ? 1 byte ? Is that an issue ?

Reply:
it should block AFAIK

Reply:
can we test it ? reading from random is blocking, but since the go documentation precises that read will read "up to" the size of the buffer, I wouldn't be surprised if it is setting O_NONBLOCK and then it could do partial reads.

If it does partial reads, we need to check the number of bytes read returned by the function, and truncate what we then pass to Write() otherwise we are passing whatever is left in the buffer after the n bytes that have been read

Reply:
I mean we could set a Deadline for the fd and check much was read but I would like to trust the kernel interface which should work if it select an existing trng.

Reply:
I see what you mean. Going do some tests

Reply:
It's O_NONBLOCK. So I fixed it